feat: Deprecates CONTENT_OBJECT_TAGS_CHANGED in favor of CONTENT_OBJECT_ASSOCIATIONS_CHANGED [FC-0062]

pomegranited opened the pull request:
**Description:** 

Deprecates the `CONTENT_OBJECT_TAGS_CHANGED` event in favor of the new `CONTENT_OBJECT_ASSOCIATIONS_CHANGED` event.

This new name and data structure better reflect the additional use of this event, which is triggered when a ContentObject's tags or [Collections change](https://github.com/openedx/edx-platform/pull/35321/files#diff-a760c3e87f2f671ac2453abbbaf6f10ca052c7acf1c46c1d36f24646e1d52a00R1231-R1235).

**ISSUE:** https://github.com/openedx/modular-learning/issues/230

Private-ref: [FAL-3787](https://tasks.opencraft.com/browse/FAL-3787)

**Dependencies:** https://github.com/openedx/openedx-events/pull/386

OK to merge this change into that PR so we can get them CC reviewed together.

**Merge deadline:** ASAP

**Installation instructions:** List any non-trivial installation instructions.

**Testing instructions:**

Unit tests should be sufficient to verify this change, but see [open-craft:edx-platform#680](https://github.com/open-craft/edx-platform/pull/680) for detailed testing instructions.

**Merge checklist:**
- [ ] All reviewers approved
- [ ] CI build is green
- [x] Version bumped
- [x] Changelog record added
- [x] Documentation updated (not only docstrings)
- [ ] Commits are squashed

**Post merge:**
- [ ] Create a tag
- [ ] Check new version is pushed to PyPI after tag-triggered build is
      finished.
- [ ] Delete working branch (if not needed anymore)

**Author concerns:** 

1. **TODO** Raise a DEPR ticket, and ensure `CONTENT_OBJECT_TAGS_CHANGED` is removed after Sumac.

pomegranited commented:
@bradenmacdonald Do we need to deprecate the old event before changing it here? It's unlikely that anyone is using `CONTENT_OBJECT_TAGS_CHANGED`, but tagging was released with Redwood, so they might be?

bradenmacdonald commented:
@pomegranited They might be in the future too (between now and whenever they upgrade to Sumac), so I think we should deprecate it. We could emit both events for a while; it doesn't really cost us anything to emit `CONTENT_OBJECT_TAGS_CHANGED` if nobody is listening to it.

pomegranited committed:
feat: adds event CONTENT_OBJECT_ASSOCIATOONS_CHANGED

and ContentObjectChangedData, which has a field for indicating what has changed.

pomegranited committed:
chore: updates changelog

rpenido reviewed (approved):
LGTM :+1:
Thank you for your work, @pomegranited!

I have nothing to note here.

- [x] I tested this using the instructions from https://github.com/open-craft/edx-platform/pull/680#pullrequestreview-2284433119
- [x] I read through the code
- [x] ~I checked for accessibility issues~
- [x] Includes documentation
